README.md
## UPD.py

Simple software to call UPD regions from germline exome/wgs trios.

## Documentation

### Installation

`pip install --editable .` or `python setup.py install`

The only dependencies are click and coloredlogs.


### Input VCF requirements

* Trio (proband-mother-father) VCF is required (i.e. all three individuals in the same VCF).
* GQ must be present in genotype fields.
* Must be annotated with some sort of population frequency. This can also be in the CSQ annotation (default: MAX_AF), if so use `--vep`


### Running

```bash
upd --vcf input.vcf.gz --proband PB_ID --mother MOTHER_ID --father FATHER_ID regions --out upd_regions.bed
```

Where PB_ID/MOTHER_ID/FATHER_ID are the sample IDs from the vcf header.

#### Optional parameters
Command |Parameter | Description
------- |--------- | -----------
base | **--min-af (DEFAULT: 0.05)** | Minimum population frequency required to include the SNP in the analysis.
base | **--af-tag (DEFAULT: MAX_AF)** | Specifies which frequency field to be used when selecting SNPs.
base | **--min-gq (DEFAULT: 30)** | Specifies the minimum GQ required to include a variant in the analysis. All three individuals' must have a GQ larged than or equal to this.
base | **--vep (flag)** | If given, search the CSQ field for `af-tag`
base | **--vep (flag)** | If given, search the CSQ field for `af-tag`
regions | **--min-sites (DEFAULT: 3)** | Minimum number of consecutive UPD sites needed to call an UPD region.
regions | **--min-size (DEFAULT: 1000)** | Minimum number of base pairs between first and last UPD site in a region required to call it.
regions/sites | **--out (DEFAULT: stdout)** | If the results should be printed to a file
regions/sites | **--iso-het-pct (DEFAULT: 0.01)** | Threshold ratio for calling homodisomy


### Output

#### Called regions (udp regions)
BED file of all called regions fulfilling the --min-sites and --min-size filter criteria, including some annotation.

Example call:
```
15      22958103        101910550       ORIGIN=PATERNAL;TYPE=HETERODISOMY;LOW_SIZE=78952446;INF_SITES=182;SNPS=2896;HET_HOM=1275/1440;OPP_SITES=0;START_LOW=20170150;END_HIGH=102516492;HIGH_SIZE=82346342
```

Field | Description
----- | -----------
ORIGIN | The parental origin of the UPD region
TYPE | Indicates whether region is heterodisomic or isodisomic/deletion. See caveats
LOW_SIZE | Lower bound of the estimated size of the called region. Distance between first and last UPD site of the region
INF_SITES | Number of UPD informative sites in the called regions
SNPS | Total number of SNPs in the region
HET_HOM | Number of heterozygous and homozygous sites in the region. Used to indicate if region has LOH
OPP_SITES | Number of sites in the region that have the opposite parental origin. So no other number than 0 have been observed...
START_LOW | Earliest possible start positon of the UPD region (position of last anti-UPD site prior to the region)
END_HIGH | Last possible end position of the UPD region (position of first anti-UPD site prior to the region)
HIGH_SIZE | Upper bound of the estimated size of the called region. Distance between the surrounding anti-UPD sites.

#### Informative sites (upd sites)
Fourth field indicates what type of field:

Type | Description
---- | -----------
ANTI_UPD | Site where proband has inherited alleles from both parents (e.g. AA x BB = AB)
UPD_MATERNAL_ORIGIN | Site where both alleles are inherited from the mother, or a fathers allele was deleted. (e.g. AA x BB = AA)
UPD_PATERNAL_ORIGIN | Site where both alleles are inherited from the father, or a mothers allele was deleted. (e.g. AA x BB = BB)
PB_HETEROZYGOUS | Heterozygous site in the proband (used only to call hetero/isodisomy)
PB_HOMOZYGOUS | Homozygous site in the proband (used only to call hetero/isodisomy)
UNINFORMATIVE | Various sites excluded from the analysis. Ignore these...


### Caveats
The isodisomic/heterodisomic calling depends on estimating if there is a run of homozygousity in the called region. This is called using presence of heterozygous sites within the call, and should only be seen as a rough estimate for smaller calls. For more statistically sound detection of this, use e.g. bcftools roh to detect regions of homozygozity and combine the results.

Isodisomies cannot be differentiated from deletions using this software. Combine the calls with SV/CNV calls to determine if there is an overlapping deletion. If overlapping deletion, the "UPD" call can be used to detemine which parental allele was deleted.






CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

About changelog [here](https://keepachangelog.com/en/1.0.0/)

## [0.1.1]
### Added
- Add argument to change the homozygozity rate cutoff for calling isodisomies
### Fixed
- Change HOMODISOMY to ISODISOMY in the output files.
### Changed


## [0.1]
First version.


tests/conftest.py
import pytest

@pytest.fixture()
def vcf_path():
    return 'tests/fixtures/test.vcf.gz'

@pytest.fixture()
def regions_output():
    line = ("15\t22958118\t101910531\tORIGIN=PATERNAL;TYPE=HETERODISOMY;LOW_SIZE=78952412;"
            "INF_SITES=138;SNPS=2032;HET_HOM=889/1006;OPP_SITES=0;START_LOW=20170126;END_HIGH="
            "102516586;HIGH_SIZE=82346460")
    return line


@pytest.fixture()
def ped_path():
    return 'tests/fixtures/test.fam'


tests/test_cli.py
from click.testing import CliRunner

from upd.cli import cli
from upd.__version__ import __version__

def test_version():
    runner = CliRunner()
    result = runner.invoke(cli, ['--version'])
    
    assert result.exit_code == 0
    assert result.output == __version__+'\n'

def test_upd(vcf_path):
    runner = CliRunner()
    result = runner.invoke(cli, [
        '--vcf', vcf_path, '--proband', 'TEST_PROBAND', '--mother', 'TEST_MOTHER', '--father', 
        'TEST_FATHER', '--vep', 'regions'
    ])
    
    assert result.exit_code == 0



tests/test_vcf_tools.py
import pytest

from upd.vcf_tools import (check_samples, get_vcf, Vcf)

def test_check_samples():
    ## GIVEN a list three samples
    sids = ['TEST_PROBAND', 'TEST_MOTHER', 'TEST_FATHER']
    proband = 'TEST_PROBAND'
    mother = 'TEST_MOTHER'
    father = 'TEST_FATHER'
    
    ## WHEN checking if the samples are in sids
    res = check_samples(sids, proband, mother, father)
    
    ## THEN assert that res is true
    assert res == True

def test_check_samples_missing_individual():
    ## GIVEN a list of two samples
    sids = ['TEST_PROBAND', 'TEST_MOTHER']
    proband = 'TEST_PROBAND'
    mother = 'TEST_MOTHER'
    father = 'TEST_FATHER'
    
    ## WHEN checking if the samples are in sids
    res = check_samples(sids, proband, mother, father)
    
    ## THEN assert that res is false
    assert res == False

def test_get_vcf(vcf_path):
    ## GIVEN the path to a VCF and the individuals that should be there
    proband = 'TEST_PROBAND'
    mother = 'TEST_MOTHER'
    father = 'TEST_FATHER'
    
    ## WHEN getting the vcf
    vcf_obj = get_vcf(vcf_path, proband, mother, father)
    
    ## THEN assert that vcf_obj has the correct type
    assert isinstance(vcf_obj, Vcf)

def test_get_nonexisting_vcf():
    ## GIVEN the path to a non existing VCF and the individuals that should be there
    proband = 'TEST_PROBAND'
    mother = 'TEST_MOTHER'
    father = 'TEST_FATHER'
    vcf_path = 'hej.vcf'
    
    ## WHEN getting the vcf
    with pytest.raises(OSError):
        ## THEN assert a non OSError is raised
        vcf_obj = get_vcf(vcf_path, proband, mother, father)

def test_get_wrong_formated_vcf(ped_path):
    ## GIVEN the path to a non existing VCF and the individuals that should be there
    proband = 'TEST_PROBAND'
    mother = 'TEST_MOTHER'
    father = 'TEST_FATHER'
    
    ## WHEN getting the vcf
    with pytest.raises(SyntaxError):
    ## THEN assert a non OSError is raised
        vcf_obj = get_vcf(ped_path, proband, mother, father)

    
def test_get_vcf_wrong_individual(vcf_path):
    ## GIVEN the path to a VCF and the individuals that should be there
    proband = 'TEST_PROBAND'
    mother = 'TEST_MOTHER'
    father = 'JOHN DOE'
    
    ## WHEN getting the vcf
    with pytest.raises(SyntaxError):
        ## THEN assert a SyntaxError is raised
        vcf_obj = get_vcf(vcf_path, proband, mother, father)
    

upd.py
#!/usr/bin/env python2
from cyvcf2 import VCF
import argparse

# Constants
UNINFORMATIVE       = 0
UPD_MATERNAL_ORIGIN = 1
UPD_PATERNAL_ORIGIN = 2
ANTI_UPD            = 3
PB_HOMOZYGOUS       = 4
PB_HETEROZYGOUS     = 5
site_type_name = ["UNINFORMATIVE", "UPD_MATERNAL_ORIGIN", "UPD_PATERNAL_ORIGIN", "ANTI_UPD", "PB_HOMOZYGOUS", "PB_HETEROZYGOUS"]

def main():

    opt = parse_arguments()

    vcf_reader = VCF(opt.vcf)

    # Check if the given samples IDs exist in the VCF header
    sids = vcf_reader.samples
    if not all(elem in sids for elem in [opt.proband, opt.mother, opt.father]):
        raise SystemExit("ERROR: At least one of the given sample IDs do not exist in the VCF header")

    # Parse the VEP CSQ header
    csq_fields = parse_CSQ_header(vcf_reader)

    # Make sure the given VEP field exists
    if opt.vep_af not in csq_fields:
        raise SystemExit("ERROR: The field {} does not exist in the VEP annotations".format(opt.vep_af))

    # Get all UPD informative sites into a list 
    site_calls = get_UPD_informative_sites(vcf_reader, opt, csq_fields, sids)

    # Make region calls
    calls = call_regions(site_calls)

    # Output BED file with results
    output_filtered_regions(opt.out, calls, opt.min_sites, opt.min_size)

    # Output individual informative sites if requested
    if opt.out_sites:
        output_informative_sites(opt.out_sites, site_calls)

def get_UPD_informative_sites(vcf, opt, csq_fields, sids):
    """Get UPD calls for each informative SNP above given pop freq"""
    site_calls = []
    for var in vcf:
    
        # Raise error if multi-allelic site
        if len(var.ALT) > 1:
            raise SystemExit('ERROR: Split your variants!')

        # Skip non-SNPs
        if not var.is_snp:
            continue
        
        # Skip variants with population frequency < threshold
        if opt.min_af > get_pop_AF(var.INFO["CSQ"], csq_fields, opt.vep_af):
            continue
        
        # Skip variants where any individual has GQ < threshold
        if not all(gq >= opt.min_gq for gq in var.gt_quals):
            continue

        # Make UPD call for site and append to call list
        proband_idx = sids.index(opt.proband)
        mother_idx  = sids.index(opt.mother)
        father_idx  = sids.index(opt.father)

        gt = var.gt_types
        pos_call = upd_site_call(gt[proband_idx], gt[mother_idx], gt[father_idx])
        site_calls.append( {'chrom':var.CHROM, 'pos':var.POS, 'call':pos_call} )
    
    return site_calls


def call_regions(sites):
    """Takes a list of UPD informative sites and generates called regions"""
    opposite = {UPD_MATERNAL_ORIGIN:UPD_PATERNAL_ORIGIN, UPD_PATERNAL_ORIGIN:UPD_MATERNAL_ORIGIN}
    calls = []
    putative_call, prev = None, None
    last_seen_anti = 0

    for c in sites:
        if not putative_call:
            # Create new putative call if UPD informative position
            if c["call"] in [UPD_MATERNAL_ORIGIN, UPD_PATERNAL_ORIGIN]:
                putative_call = {
                    'call':c['call'],
                    'chrom':c['chrom'],
                    'start_lo':last_seen_anti['pos'],
                    'start_hi':c['pos'],
                    'end_lo':c['pos'],
                    'end_hi':c['pos'],
                    'run_len': 1,
                    'opposites': 0,
                    'hom_sites': 1,
                    'het_sites': 0,
                    'tot': 1
                }

            # Save last positive which is definately not in UPD region
            if c["call"] == ANTI_UPD or not prev or c['chrom'] != prev['chrom']:
                last_seen_anti = {'chrom':c['chrom'], 'pos':c['pos']}

        else:

            # If an anti-UPD site or end of chromosome => close the putative call
            if c["call"] == ANTI_UPD or c['chrom'] != putative_call['chrom']:
                if c["call"] == ANTI_UPD:
                    putative_call['end_hi'] = c['pos']-1
                else:
                    putative_call['end_hi'] = prev['pos']
                calls.append(putative_call)
                last_seen_anti = {'chrom': c['chrom'], 'pos': c['pos']}      
                putative_call = None

            else:
                # If site call is same as putative call => extend the putative region
                if c['call'] == putative_call['call']:
                    putative_call['end_lo'] = c['pos']
                    putative_call['run_len'] += 1

                if c['call'] == PB_HOMOZYGOUS:
                    putative_call['hom_sites'] += 1
                if c['call'] == PB_HETEROZYGOUS:
                    putative_call['het_sites'] += 1
                    
                # If site call is opposite (maternal<->paternal), count it. (This pretty much never happens?)
                if c['call'] == opposite[putative_call['call']]:
                    putative_call['opposites'] += 1

                # Count total number of SNPs in the call region
                putative_call['tot'] += 1

        prev = c

    return calls


def output_filtered_regions(outfile, calls, min_sites, min_size):
    """Takes called regions, filters them, and outputs annotated BED file"""
    with open(outfile, 'w') as out:
    
        for rcall in calls:
            if rcall['run_len'] > min_sites:
                call_str = {2: 'PATERNAL', 1: 'MATERNAL'}
                lo_size = rcall['end_lo'] - rcall['start_hi']
                hi_size = rcall['end_hi'] - rcall['start_lo']

                # Rough estimation if heterodisomy or homodisomy/deletion
                het_pct = float(rcall['het_sites']) / float((rcall['het_sites']+rcall['hom_sites']))
                upd_type = "HETERODISOMY"
                if het_pct < 0.01:
                    upd_type = "HOMODISOMY/DELETION"
                if lo_size >= min_size:
                    out.write("{}\t{}\t{}\tORIGIN={};TYPE={};LOW_SIZE={};INF_SITES={};SNPS={};HET_HOM={}/{};OPP_SITES={};START_LOW={};END_HIGH={};HIGH_SIZE={}\n".format(
                        rcall['chrom'],
                        rcall['start_hi']-1,
                        rcall['end_lo'],
                        call_str[rcall['call']],
                        upd_type,
                        lo_size,
                        rcall['run_len'],
                        rcall['tot'],
                        rcall['het_sites'],
                        rcall['hom_sites'],
                        rcall['opposites'],
                        rcall['start_lo'],
                        rcall['end_hi'],
                        hi_size
                    ))



def output_informative_sites(outfile, calls):
    """Outputs all called informative sites to a BED file"""
    with open(outfile, 'w') as out:
    
        for scall in calls:
            
            out.write("{}\t{}\t{}\t{}\n".format(
                scall['chrom'],
                scall['pos']-1,
                scall['pos'],
                site_type_name[ scall['call'] ]
            ))
                    

def upd_site_call(gt_pb, gt_mo, gt_fa):
    """ Call UPD informative sites, input genotype coding: 0=HOM_REF, 1=HET, 3=HOM_ALT, 2=other """

    opposite = {0:3, 3:0}
    
    if gt_pb == 2 or gt_mo == 2 or gt_fa == 2:
        return UNINFORMATIVE

    if gt_pb == 0 or gt_pb == 3:
        opp = opposite[gt_pb]
        if gt_mo == opp and gt_fa == opp:
            return UNINFORMATIVE
        if gt_mo == opp:
            return UPD_PATERNAL_ORIGIN
        if gt_fa == opp:
            return UPD_MATERNAL_ORIGIN
        return PB_HOMOZYGOUS
        
    if gt_pb == 1:
        if gt_mo != 1 and gt_fa != 1 and gt_mo == opposite[gt_fa]:
            return ANTI_UPD
        return PB_HETEROZYGOUS

    return UNINFORMATIVE


def get_header_desc(reader, header_id):
    """ Get description field of an header field ID """
    for rec in reader.header_iter():
        d = rec.info()
        if d.get('ID') == header_id:
            return d.get('Description')

    return None


def parse_CSQ_header(reader):
    """Parse the order of VEP fields from the vcf header"""
    csq_str = get_header_desc(reader, "CSQ")

    if csq_str:
        _, csq_format_str = csq_str.split('Format: ')
        csq_format_str = csq_format_str.rstrip('"')
        csq_format = csq_format_str.split('|')
        return csq_format

    else:
        raise ValueError("CSQ header field missing. The VCF need to be annotated with VEP")

    
def get_pop_AF(vep_data, vep_fields, AF_str):
    """Extract population frequency from VEP annotations, returns 0 if no data."""

    first_vep_str = vep_data.split(',')[0]
    data = first_vep_str.split('|')

    for i in range(len(data)):
        if vep_fields[i] == AF_str:
            return float(data[i] or 0)

    return 0


def parse_arguments():
    parser = argparse.ArgumentParser(description='Aggregate vcfs from different variant callers.')
    parser.add_argument('vcf', metavar='vcf', type=str, help='input vcf file')
    parser.add_argument('--proband', type=str, help='ID of proband in VCF', required=True)
    parser.add_argument('--mother', type=str, help='ID of mother in VCF', required=True)
    parser.add_argument('--father', type=str, help='ID of father in VCF', required=True)
    parser.add_argument('--vep_af', type=str, default='MAX_AF', help='Which VEP fields to use for poulation frequency filtering')
    parser.add_argument('--min_af', type=float, default=0.05, help='Minimum SNP frequency')
    parser.add_argument('--min_gq', type=int, default=30, help='Minimum GQ score')
    parser.add_argument('--min_sites', type=int, default=3, help='Minimum UPD informative sites required to call a region')
    parser.add_argument('--min_size', type=int, default=1000, help='Minimum size (bp) required to call a region')
    parser.add_argument('--out', type=str, required=True, help='Output bed file of called regions')
    parser.add_argument('--out_sites', type=str, default=None, help='Output bed file of all informative sites')

    opt = parser.parse_args()
    return opt


if __name__ == '__main__':
    main()


upd/utils.py
import logging

from .vcf_tools import get_pop_AF

LOG = logging.getLogger(__name__)

UNINFORMATIVE       = 0
UPD_MATERNAL_ORIGIN = 1
UPD_PATERNAL_ORIGIN = 2
ANTI_UPD            = 3
PB_HOMOZYGOUS       = 4
PB_HETEROZYGOUS     = 5


def upd_site_call(gt_pb, gt_mo, gt_fa):
    """Call UPD informative sites
    
    input genotype coding: 0=HOM_REF, 1=HET, 3=HOM_ALT, 2=other

    Args:
        gt_pb (int): Genotype of proband as integer
        gt_mo (int): Genotype of mother as integer
        gt_fa (int): Genotype of father as integer
    
    Returns:
        site_info (int): One of the globals
    
    """

    opposite = {0:3, 3:0}
    # If any of the individuals are 'other' the site is UNINFORMATIVE
    if gt_pb == 2 or gt_mo == 2 or gt_fa == 2:
        return 0

    if gt_pb == 0 or gt_pb == 3:
        opp = opposite[gt_pb]
        if gt_mo == opp and gt_fa == opp:
            return UNINFORMATIVE
        if gt_mo == opp:
            LOG.debug("Found paternal upd")
            return UPD_PATERNAL_ORIGIN
        if gt_fa == opp:
            LOG.debug("Found maternal upd")
            return UPD_MATERNAL_ORIGIN
        return PB_HOMOZYGOUS
        
    if gt_pb == 1:
        if gt_mo != 1 and gt_fa != 1 and gt_mo == opposite[gt_fa]:
            LOG.debug("Found anti upd")
            return ANTI_UPD
        return PB_HETEROZYGOUS

    return UNINFORMATIVE


def get_UPD_informative_sites(vcf, csq_fields, proband, mother, father, min_af=0.05, 
                              af_tag='MAX_AF', min_gq=30):
    """Get UPD calls for each informative SNP above given pop freq
    
    Args:
        vcf (upd.vcf_tools.Vcf)
        min_af (float): Minimum allele frequency to consider SNP
        af_tag (str): Key to AF in annotation
        min_gq (int): Minimum GQ to consider variant
        csq_fields (list): describes VEP annotation
        sids (list): describes what position inds have in VCF
        proband (str): ID of proband in VCF
        mother (str): ID of mother in VCF
        father (str): ID of father in VCF
        
    Yields:
        site_calls (dict): A generator with dictionaries that describes the variant.

    """
    site_calls = []
    # Make UPD call for site and append to call list
    sids = vcf.samples
    proband_idx = sids.index(proband)
    mother_idx  = sids.index(mother)
    father_idx  = sids.index(father)
    
    nr_informative = 0
    
    for i,var in enumerate(vcf,1):
    
        # Raise error if multi-allelic site
        if len(var.ALT) > 1:
            raise SystemExit('ERROR: Split your variants!')

        # Skip non-SNPs
        if not var.is_snp:
            continue
        
        # Skip variants with population frequency < threshold
        if min_af > get_pop_AF(var, csq_fields, af_tag):
            continue
        
        # Skip variants where any individual has GQ < threshold
        if not all(gq >= min_gq for gq in var.gt_quals):
            continue

        gt = var.gt_types
        pos_call = upd_site_call(gt[proband_idx], gt[mother_idx], gt[father_idx])
        
        nr_informative += 1
        yield {'chrom':var.CHROM, 'pos':var.POS, 'call':pos_call}
    
    LOG.info("%s variants in vcf", i)
    LOG.info("%s informative variants found", nr_informative)

def call_regions(sites):
    """Yields called regions
    
    A putative_call is a collection of information from the variants in a upd region. 
    The call is started if there are variants that indicates UPD maternal or paternal origin. When 
    a anti upd site is encountered or the chromosome ends the call returned.
    
    Args:
        sites (iterable(dict))

    Yields:
        putative_call (dict): UPD informative region 
    """
    opposite = {UPD_MATERNAL_ORIGIN:UPD_PATERNAL_ORIGIN, UPD_PATERNAL_ORIGIN:UPD_MATERNAL_ORIGIN}
    calls = []
    putative_call = None
    prev = None
    last_seen_anti = {'chrom':'0', 'pos':0}
    
    for c in sites:
        if prev and (c['chrom'] != prev['chrom']):
            LOG.info("Chromosome %s checked", prev['chrom'])
        if not putative_call:
            # Create new putative call if UPD informative position
            if c["call"] in [UPD_MATERNAL_ORIGIN, UPD_PATERNAL_ORIGIN]:
                putative_call = {
                    'call':c['call'],
                    'chrom':c['chrom'],
                    'start_lo':last_seen_anti['pos'],
                    'start_hi':c['pos'],
                    'end_lo':c['pos'],
                    'end_hi':c['pos'],
                    'run_len': 1,
                    'opposites': 0,
                    'hom_sites': 1,
                    'het_sites': 0,
                    'tot': 1
                }

            # Save last positive which is definately not in UPD region
            if c["call"] == ANTI_UPD or not prev or c['chrom'] != prev['chrom']:
                last_seen_anti = {'chrom':c['chrom'], 'pos':c['pos']}

            prev = c
            continue

        # If an anti-UPD site or end of chromosome => close the putative call
        if c["call"] == ANTI_UPD or c['chrom'] != putative_call['chrom']:
            if c["call"] == ANTI_UPD:
                putative_call['end_hi'] = c['pos']-1
            else:
                putative_call['end_hi'] = prev['pos']
            
            yield putative_call
            
            last_seen_anti = {'chrom': c['chrom'], 'pos': c['pos']}      
            putative_call = None
            prev = c
            continue

        # If site call is same as putative call => extend the putative region
        if c['call'] == putative_call['call']:
            putative_call['end_lo'] = c['pos']
            putative_call['run_len'] += 1

        if c['call'] == PB_HOMOZYGOUS:
            putative_call['hom_sites'] += 1
        
        if c['call'] == PB_HETEROZYGOUS:
            putative_call['het_sites'] += 1
                    
        # If site call is opposite (maternal<->paternal), count it. (This pretty much never happens?)
        if c['call'] == opposite[putative_call['call']]:
            putative_call['opposites'] += 1

        # Count total number of SNPs in the call region
        putative_call['tot'] += 1

        prev = c

    LOG.info("Chromosome %s checked", prev['chrom'])

    if putative_call:
        yield putative_call

upd/bed_utils.py

def output_filtered_regions(calls, min_sites=3, min_size=1000, iso_het_pct=0.01):
    """Takes called regions, filters them, and yields annotated BED lines

    Args:
        calls (iterable): An iterable with UPD regions
        min_sites (int): Minimum UPD informative sites required to call a region
        min_size (int): Minimum size (bp) required to call a region
        iso_het_pct(float): Quota between het-sites and hom-sties
    Yields:
        out_line (str): A formated string with the information about a region
    """
    out_line = ("{}\t{}\t{}\tORIGIN={};TYPE={};LOW_SIZE={};INF_SITES={};SNPS={};HET_HOM={}/{}"
               ";OPP_SITES={};START_LOW={};END_HIGH={};HIGH_SIZE={}")

    for rcall in calls:
        if not rcall['run_len'] > min_sites:
            continue
        call_str = {2: 'PATERNAL', 1: 'MATERNAL'}
        lo_size = rcall['end_lo'] - rcall['start_hi']
        hi_size = rcall['end_hi'] - rcall['start_lo']

        # Rough estimation if heterodisomy or homodisomy/deletion
        het_pct = float(rcall['het_sites']) / float((rcall['het_sites']+rcall['hom_sites']))
        upd_type = "HETERODISOMY"
        if het_pct < iso_het_pct:
            upd_type = "ISODISOMY/DELETION"

        if lo_size < min_size:
            continue

        yield out_line.format(
                rcall['chrom'],
                rcall['start_hi']-1,
                rcall['end_lo'],
                call_str[rcall['call']],
                upd_type,
                lo_size,
                rcall['run_len'],
                rcall['tot'],
                rcall['het_sites'],
                rcall['hom_sites'],
                rcall['opposites'],
                rcall['start_lo'],
                rcall['end_hi'],
                hi_size
            )


upd/cli.py
import logging

import coloredlogs
import click
import datetime

from pprint import pprint as pp

from upd.__version__ import __version__
from upd.vcf_tools import (parse_CSQ_header, get_vcf)
from upd.utils import (get_UPD_informative_sites, call_regions)
from upd.bed_utils import (output_filtered_regions)

LOG = logging.getLogger(__name__)

LOG_LEVELS = ['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL']


def print_version(ctx, param, value):
    if not value or ctx.resilient_parsing:
        return
    click.echo(__version__)
    ctx.exit()

@click.group()
@click.option('--vcf',
    type=click.Path(exists=True),
    required=True,
)
@click.option('--proband',
    help="ID of proband in VCF",
    required=True,
)
@click.option('--mother',
    help="ID of mother in VCF",
    required=True,
)
@click.option('--father',
    help="ID of father in VCF",
    required=True,
)
@click.option('--af-tag',
    help="Which field to use for population frequency filtering",
    default='MAX_AF',
    show_default=True
)
@click.option('--vep',
    help="If af-tag is in VEP annotation",
    is_flag=True,
)
@click.option('--min-af',
    help="Minimum SNP frequency",
    default=0.05,
    show_default=True
)
@click.option('--min-gq',
    help="Minimum GQ score",
    default=30,
    show_default=True
)
@click.option('--loglevel',
    default='INFO',
    type=click.Choice(LOG_LEVELS),
    help="Set the level of log output.",
    show_default=True
)
@click.option('--version',
    is_flag=True,
    callback=print_version,
    expose_value=False,
    is_eager=True
)

@click.pass_context
def cli(context, vcf, proband, mother, father, af_tag, vep, min_af, min_gq, loglevel):
    """Simple software to call UPD regions from germline exome/wgs trios"""
    coloredlogs.install(level=loglevel)
    LOG.info("Running upd version %s", __version__)

    context.obj = {}
    context.obj['start_time'] = datetime.datetime.now()
    # Check if the given samples IDs exist in the VCF header
    try:
        vcf_reader = get_vcf(vcf, proband, mother, father)
    except Exception as err:
        LOG.warning(err)
        context.abort()

    csq_fields = None
    if vep:
        try:
            csq_fields = parse_CSQ_header(vcf_reader)
        except Exception as err:
            LOG.warning(err)
            context.abort()

        # Make sure the given VEP field exists
        if af_tag not in csq_fields:
            LOG.warning("The field %s does not exist in the VEP annotations", af_tag)
            LOG.info("Existing CSQ fields {}".format('|'.join(csq_fields)))
            context.abort()

    else:
        if not vcf_reader.contains(af_tag):
            LOG.warning("The field %s does not exist in the VCF", af_tag)
            context.abort()

    # Get all UPD informative sites into a list
    context.obj['site_calls'] = get_UPD_informative_sites(
        vcf=vcf_reader,
        csq_fields=csq_fields,
        proband=proband,
        mother=mother,
        father=father,
        min_af=min_af,
        af_tag=af_tag,
        min_gq=min_gq
    )

@cli.command()
@click.option('--min-sites',
    help="Minimum UPD informative sites required to call a region",
    default=3,
    show_default=True
)
@click.option('--min-size',
    help="Minimum size (bp) required to call a region",
    default=1000,
    show_default=True
)
@click.option('-o','--out',
    help="Output bed file of all informative sites",
    type=click.Path(exists=False),
    default='-',
)

@click.option('--iso-het-pct',
    help="Ratio iso/het for determening UPD type",
    default=0.01,
    show_default=True
)

@click.pass_context
def regions(context, min_sites, min_size, iso_het_pct, out):
    """Call UPD regions"""
    # Make region calls
    calls = call_regions(context.obj['site_calls'])

    out_lines = output_filtered_regions(calls, min_sites, min_size, iso_het_pct)

    with click.open_file(out, 'w') as f:
        for line in out_lines:
            f.write(line+'\n')

    end_time = datetime.datetime.now() - context.obj['start_time']
    LOG.info(f"Time to parse variants {end_time}")


@cli.command()
@click.option('-o','--out',
    help="Output bed file of all informative sites",
    type=click.Path(exists=False),
    default='-',
)
@click.pass_context
def sites(context, out):
    """Prints the sites that are informative for UPD"""
    site_type_name = [
        "UNINFORMATIVE", "UPD_MATERNAL_ORIGIN", "UPD_PATERNAL_ORIGIN", "ANTI_UPD",
        "PB_HOMOZYGOUS", "PB_HETEROZYGOUS"
    ]

    with click.open_file(out, 'w') as f:
        for scall in context.obj['site_calls']:
            f.write("{}\t{}\t{}\t{}\n".format(
                scall['chrom'],
                scall['pos']-1,
                scall['pos'],
                site_type_name[scall['call']]
            ))

    end_time = datetime.datetime.now() - context.obj['start_time']
    LOG.info(f"Time to parse variants {end_time}")
